### PR TITLE
placement: return rules after clone

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve?

Fix https://github.com/tikv/pd/issues/3886

### What is changed and how it works?

Clone the rule before returning to callers. It should be safe to modify and update again.

### Check List

Tests
- Unit test
- Integration test

Related changes
- Need to cherry-pick to the release branch

### Release note

```release-note
Fix an issue where data is not stored when using max-replicas or location-labels to indirectly update default placement rule
```

### CODE REVIEW

**Comment by disksing** [comment]
Please take a look. @HunDunDM @bufferflies

**Review by nolouch — APPROVED** [review]
lgtm

File: server/server.go [hunk 1]
```diff
@@ -1378,3 +1376,15 @@ func (s *Server) SaveTTLConfig(data map[string]interface{}, ttl time.Duration) e
 	}
 	return nil
 }
+
+func stringsEqual(a, b []string) bool {
```
Comment (HunDunDM) [inline server/server.go:1380]: Judge the situation where `len` is all 0, then we can still use `reflect.DeepEqual` directly?

**Review by HunDunDM — COMMENTED** [review]
Does Copy-on-Write perform better?

File: server/server.go (hunk 1 above)
Comment (bufferflies) [inline server/server.go:1380]: it will be better  if move the function to typeutil.

**Comment by disksing** [comment]
> Does Copy-on-Write perform better?

copy-on-write should be better. Since go does not have `const`, it is easy to miss use. I think for the methods that are mainly used by HTTP API, it is acceptable to clone every rule.

File: server/server.go (hunk 1 above)
Comment (disksing) [reply server/server.go:1380]: ok.

Comment (disksing) [reply server/server.go:1380]: Then it will be `if len(a) == 0 && len(b) == 0 || reflect.DeepEqual(a, )`, looks not very elegant I think.

**Comment by bufferflies** [comment]
lgtm

File: pkg/typeutil/comparison.go [hunk 2]
```diff
@@ -38,3 +38,16 @@ func MinDuration(a, b time.Duration) time.Duration {
 	}
 	return b
 }
+
+// StringsEqual checks if two string slices are equal. Empyt slice and nil are considered equal.
+func StringsEqual(a, b []string) bool {
```
Comment (HunDunDM) [inline pkg/typeutil/comparison.go:43]: Do we need the unit test?

Comment (disksing) [reply pkg/typeutil/comparison.go:43]: Too simple to test. The code was written by AI, which I feel is highly credible.

**Comment by nolouch** [comment]
PTAL @HunDunDM @bufferflies

**Comment by bufferflies** [comment]
lgtm

**Review by bufferflies — APPROVED** [review]

**Pull request merged**